src/psyclone/tests/psyir/tools/dependency_tools_test.py

```diff
@@ -202,116 +228,340 @@ def test_arrays_parallelise(parser):
+            "condition" in str(msg))
+    assert msg.code == DACode.ERROR_WRITE_WRITE_RACE
+    assert msg.var_names == ["mask(jk,jk)"]
 
     jj_symbol = loops.scope.symbol_table.lookup("jj")
     # Write to array that does not depend on parallel loop variable
     parallel = dep_tools.can_loop_be_parallelised(loops[1], jj_symbol)
     assert parallel is True
+    # pylint: disable=use-implicit-booleaness-not-comparison
     assert dep_tools.get_all_messages() == []
 
     # Use parallel loop variable in more than one dimension
     parallel = dep_tools.can_loop_be_parallelised(loops[2], jj_symbol)
-    assert parallel is False
-    assert ("Variable 'mask' is written to and the loop variable 'jj' is "
-            "used in different index locations: mask(jj,jj) and mask(jj,jj)"
-            in dep_tools.get_all_messages()[0])
+    assert parallel is True
 
     # Use a stencil access (with write), which prevents parallelisation
     parallel = dep_tools.can_loop_be_parallelised(loops[3], jj_symbol)
     assert parallel is False
-    assert ("Variable 'mask' is written and is accessed using indices "
-            "'jj + 1' and 'jj' and can therefore not be parallelised"
-            in dep_tools.get_all_messages()[0])
+    msg = dep_tools.get_all_messages()[0]
+    assert ("The write access to 'mask(ji,jj)' and to 'mask(ji,jj + 1)' are "
+            "dependent and cannot be parallelised."
+            in str(msg))
+    assert msg.code == DACode.ERROR_DEPENDENCY
+    assert msg.var_names == ["mask(ji,jj)", "mask(ji,jj + 1)"]
+
+
+# -----------------------------------------------------------------------------
+@pytest.mark.parametrize("expression, correct",
+                         [("a1(i+i+j)", [set(("i", "j"))]),
+                          ("a1(1)", [set()]),
+                          ("a2(i+j,2*j+k+1)", [set(("i", "j")),
+                                               set(("j", "k"))]),
+                          ("a3(i,j,i)", [set("i"), set("j"), set("i")]),
+                          ("dv(i)%a(j)%b(k)", [set("i"), set("j"),
+                                               set("k")])])
+def test_get_indices(expression, correct, parser):
+    '''Tests that getting the indices of an array expressions
+    works as expected.
+    '''
+    reader = FortranStringReader(f'''program test
+                                 use my_mod, only: my_type
+                                 type(my_type) :: dv(10)
+                                 integer i, j, k
+                                 integer, parameter :: n=10
+                                 real, dimension(n) :: a1
+                                 real, dimension(n,n) :: a2
+                                 real, dimension(n,n,n) :: a3
+                                 {expression} = 1
+                                 end program test''')
+    prog = parser(reader)
+    psy = PSyFactory("nemo", distributed_memory=False).create(prog)
+    assign = psy.invokes.get("test").schedule
```

Comment:
Is there any reason for many of the DependencyTool tests to need the FortranStringReader and PSyFactory?

Ideally they should work with the more generic `psyir = fortran_reader.psyir_from_source(source)`